cuqi/solver/_solver.py

```diff
@@ -320,6 +326,137 @@ def solve(self):
+        else:
+            self.explicitPinv = False
+            Pinv = None # we do cholesky.solve or sparse.solve with P
+            if has_cholmod:
+                # turn P into a cholesky object
+                P = cholesky(P, ordering_method='natural')
+        self.Pinv = Pinv # inverse of the preconditioner as a ndarray or function
+
+    def solve(self):
+        # initial state
+        x = self.x0.copy()
+        if self.explicitA:
+            r = self.b - (self.A @ x)
+            if self.explicitPinv:
+                s = self.Pinv.T @ (self.A.T @ r)
+            else:
+                s = self._apply_Pinv((self.A.T @ r), 2)
+        else:
+            r = self.b - self.A(x, 1)
+            if self.explicitPinv:
+                s = self.Pinv.T @ self.A(r, 2)
+            else:
+                s = self._apply_Pinv(self.A(r, 2), 2)
+
+        # initialization
+        p = s.copy()
+        norms0 = LA.norm(s)
+        normx = LA.norm(x)
+        gamma, xmax = norms0**2, normx
+    
+        # main loop
+        k, flag, indefinite = 0, 0, 0
+        while (k < self.maxit) and (flag == 0):
+            k += 1
+            #
+            if self.explicitPinv:
+                t = self.Pinv @ p
+            else:
+                t = self._apply_Pinv(p, 1)
+            if self.explicitA:
+                q = self.A @ t
+            else:
+                q = self.A(t, 1)
+            #
+            delta_cgls = LA.norm(q)**2
+            if (delta_cgls < 0):
+                indefinite = True
+            elif (delta_cgls == 0):
+                delta_cgls = eps
+            alpha_cgls = gamma / delta_cgls
+            #
+            x += alpha_cgls*t
+            r -= alpha_cgls*q
+            if self.explicitA:
+                if self.explicitPinv:
+                    s = self.Pinv.T @ (self.A.T @ r)
+                else:
+                    s = self._apply_Pinv((self.A.T @ r), 2)
+            else:
+                if self.explicitPinv:
+                    s = self.Pinv.T @ self.A(r, 2) 
+                else:
+                    s = self._apply_Pinv(self.A(r, 2), 2)
+            #
+            norms = LA.norm(s)
+            gamma1 = gamma.copy()
+            gamma = norms**2
+            beta = gamma / gamma1
+            p = s + beta*p
+
+            # convergence
+            normx = LA.norm(x)
+            xmax = max(xmax, normx)
+            flag = (norms <= norms0*self.tol) or (normx*self.tol >= 1)
+            # resNE = norms / norms0
+
+        shrink = normx/xmax
+        if indefinite:          
+            flag = 3   # Matrix (A'*A + delta*L) seems to be singular or indefinite
+            ValueError('\n Negative curvature detected !')  
+        if shrink <= np.sqrt(self.tol):
+            flag = 4   # Instability likely: (A'*A + delta*L) indefinite and NORM(X) decreased
+            ValueError('\n Instability likely !') 
+
+        return x, k
+
+    def _apply_Pinv(self, x, flag):
+        if has_cholmod:
```

Comment:
For future reference could you please add a brief description in the docstring of this method what it does?